test/upgrade/upgrade.go

```diff
@@ -42,3 +51,251 @@ func RunMainTest(m *testing.M) {
 		return m.Run()
 	}())
 }
+
+// DurableFeature holds the setup and verify phase of a feature. The "setup" phase should set up
+// the feature. The "verify" phase should only verify its function. This function should be
+// idempotent. Calling this function multiple times should still properly verify the feature
+// (e.g. one call after upgrade, one call after downgrade).
+type DurableFeature struct {
```

Comment:
This could also be named `FeatureWithEnvironment` but DurableFeature might be easier to remember.